# MRD-Adaptis
## ABSTRACT
Detection of variations through circulating tumor DNA (ctDNA) has become a key method for monitoring minimal residual disease (MRD). However, fluctuations in detection performance across samples have emerged as a new challenge, as ctDNA samples exhibit significant heterogeneity in the limit of detection (LOD) and in SV types. SV callers are equipped with multiple user-configurable parameters designed to accommodate sample heterogeneity, but optimizing these parameters is difficult for users. Thus, we propose a self-adaptive mechanism that optimizes parameters using features derived from ctDNA data. This presents a particularly challenging problem because the underlying associations between parameters and features are difficult to discern, and the optimization problem itself is NP-hard. To overcome these challenges, we train a meta-model using metadata generated by extracting meta-features and optimizing parameters within a Bayesian optimization framework from window-segmented heterogeneous sequencing samples. This approach transforms the traditionally NP-hard problem, which typically requires exhaustive search or complex optimization strategies, into a manageable learning task. This approach enables the rapid recommendation of optimal parameter settings for each candidate region based on their meta-features, ensuring stable and sensitive detection at the sample level. We conducted extensive validation experiments on simulated data and three real-world ctDNA datasets. For each dataset, we evaluated precision, recall, and F1-score for every sample and performed comprehensive statistical analyses of these metrics, including mean, standard deviation (std), coefficient of variation (CV), root mean square error (RMSE), kurtosis, and skewness. MRD-Adaptis consistently outperformed existing methods across all metrics, highlighting its enhanced stability and reliability in SV detection.
<img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure1.png" alt="Figure 1" width="600">


## Usage
### Simulated Data files Generation

Simulate fq data using GSDcreator by following the method described in the paper available on IEEE Xplore.

**Reference:** 
[Simulation method for fq data](https://ieeexplore.ieee.org/abstract/document/8983192)

**Requirements:**
- Python version 2.7
- **Steps to Generate Simulated Data:**

1. **Generate Simulation Scripts**
   Use `A_stableCallerPaperSimFlow.py` to generate `sh` scripts for simulating 10,000 samples:

   ```bash
   python A_stableCallerPaperSimFlow.py -o sh_files
   ```
### Simulated Data Generation with GSDcreator

Generate simulation scripts for 10,000 samples, install GSDcreator, and run the simulations as follows:
**Requirements:**
- Python version 2.7
1. **Generate Simulation Scripts:**
   To create simulation scripts for 10,000 samples, use the following command:

2. **Install GSDcreator:**
Install GSDcreator and import the `sh_files`.

   ```bash
    dos2unix 10000run.py
    chmod +x 10000run.py
   ```
4. **Prepare the Simulation Script:**
Convert the script to Unix format and make it executable:

5. **Run the Simulation:**
Ensure Python version 2.7 then execute the script:
    ```bash
    python 10000run.py
    ```
6.**To distribute the converted `.vcf` files into each simulation folderï¼š
Move your `shinvcf.py` script to the Python 2.7 environment's `bin` directory:**
   ```bash
     mv noinsertshinvcf.py /yourpath/py2env/bin/
     chmod +x /home/cloudam/.conda/envs/py2env/bin/shinvcf.py
     dos2unix /home/cloudam/.conda/envs/py2env/bin/shinvcf.py
     shinvcf.py A_stableCallerPaperSimFlowShell.sh base.vcf
```
7.To split the samples into segments ranging from thousands (kilobases) to millions (megabases) of base pairs, use the following approach:
- Python version 3.6 or higher.Splitting Samples with Multithreading Support:
   ```bash
   python split_bv32.py
   ```
<img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure2.png" alt="Figure 2" width="600">
### Extracting Sample Meta-Features
- Python version 3.6 or higher
   ```bash
   python aAT0047/MRD-Adaptis/metafeature.py
   ```
   
### Initializing  opt Parameter Configuration  with Multi-Threading- Python version 3.6 or higher
 ```bash
   
   python aAT0047/MRD-Adaptis/main.py
   ```
<img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure3.png" alt="Figure 3" width="600">
###  Training a meta-model, in the context of machine learning, involves creating a model that can learn from the outputs or the performance of other models.then get multi_target_regression_model.pth
 ```bash
   
   python aAT0047/MRD-Adaptis/metaleaner.py
   ```
<img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure4.png" alt="Figure 4" width="600">
### Testing model with recommended parameters 
```bash
   
   python model.py 1.bam
   <img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure5.png" alt="Figure 5" width="600">
   print
    prediction_dict = {
         "w"
        "msw": ,
        "tt": ,
        "back_distance": ,
        "min_mapping_threshold":],
        "min_clip":,
        "read_length":,
        "min_non_overlap":,
        "discordant_z": 
    }
   ```
<img src="https://github.com/aAT0047/MRD-Adaptis/blob/main/image/figure6.png" alt="Figure 6" width="600">
### Testing DELLY, LUMPY, Manta, BreakDancer, Pindel, MetaSV, SvABA, and SVstabilizer
 ```bash
   
   python /SVfolder/vsworkflow/callerworkflow.py
   ```


